### PR TITLE
Fix DeepMIL metrics bug

### PR DESCRIPTION
**Pull request by dccastro**
> PR in draft mode pending merging of https://github.com/microsoft/hi-ml/pull/195 to update the submodule.

Fixes a bug in DeepMILModule whereby predicted hard labels were being passed to classification metrics, instead of the intended continuous probability scores. This PR also adds tests to ensure this won't happen anymore.

### CODE REVIEW

**Review by maxilse — DISMISSED** [review]

**Review by maxilse — APPROVED** [review]

File: InnerEye/ML/Histopathology/models/deepmil.py [hunk 1]
```diff
@@ -179,12 +179,13 @@ def get_metrics(self) -> nn.ModuleDict:
...
+            return nn.ModuleDict({MetricsKey.ACC: Accuracy(threshold=threshold),
                                   MetricsKey.AUROC: AUROC(num_classes=self.n_classes),
-                                  MetricsKey.PRECISION: Precision(),
-                                  MetricsKey.RECALL: Recall(),
-                                  MetricsKey.F1: F1(),
-                                  MetricsKey.CONF_MATRIX: ConfusionMatrix(num_classes=self.n_classes+1)})
+                                  MetricsKey.PRECISION: Precision(threshold=threshold),
+                                  MetricsKey.RECALL: Recall(threshold=threshold),
+                                  MetricsKey.F1: F1(threshold=threshold),
+                                  MetricsKey.CONF_MATRIX: ConfusionMatrix(num_classes=2, threshold=threshold)})
```
Comment (vale-salvatelli) [inline InnerEye/ML/Histopathology/models/deepmil.py:188]: before it was `num_classes=self.n_classes + 1`, was this change intentional?

File: environment.yml [hunk 2]
```diff
@@ -34,7 +34,7 @@ dependencies:
       - jupyter-client==6.1.5
       - lightning-bolts==0.4.0
       - matplotlib==3.3.0
-      - mlflow==1.17.0
+      - mlflow==1.23.1
```
Comment (vale-salvatelli) [inline environment.yml:37]: why is this needed?

**Comment by vale-salvatelli** [comment]
For my own understanding - why the hi-ml update is needed for this PR? 
I saw the PR in hi-ml to add the same change in there but I guess the two things go in parallel. Am I missing other changes?

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 1 above)
Comment (dccastro) [reply InnerEye/ML/Histopathology/models/deepmil.py:188]: This block is executed only when `n_classes == 1` (i.e. binary) so I preferred to make this explicit.

File: environment.yml (hunk 2 above)
Comment (dccastro) [reply environment.yml:37]: Component Governance had a high-severity alert for `mlflow`.

**Comment by dccastro** [comment]
No other crucial changes in hi-ml related to this PR, but thought it prudent to keep it up-to-date. I can revert it if you prefer.

File: Tests/ML/histopathology/models/test_deepmil.py [hunk 3]
```diff
@@ -157,6 +160,106 @@ def test_lightningmodule_mean_pooling(
...
+                             for key in thresholded_metrics_keys}
+
+    reset_metrics(module_metrics_dict)
+    set_metrics_threshold(module_metrics_dict, threshold=high_threshold)
+    _ = module.test_step(batch, 0)
+    results_high_threshold = {key: module_metrics_dict[key].compute()
+                              for key in thresholded_metrics_keys}
+
+    for key in thresholded_metrics_keys:
+        assert not torch.allclose(results_low_threshold[key], results_high_threshold[key]), \
```
Comment (vale-salvatelli) [inline Tests/ML/histopathology/models/test_deepmil.py:259]: the default tolerance is quite low, can make sense to pass it a larger number given here we're interested in checking these results ARE different?

File: InnerEye/ML/Histopathology/models/deepmil.py (hunk 1 above)
Comment (vale-salvatelli) [reply InnerEye/ML/Histopathology/models/deepmil.py:188]: I see now. I think the fact self.n_classes=1 and num_classes=2 are equivalent cases is a bit confusing. We might want to revise this for clarity at a certain point

File: environment.yml (hunk 2 above)
Comment (vale-salvatelli) [reply environment.yml:37]: OK - I think these are the changes we need to keep under control when we finally decide to decouple the histo env. It can easily drag a lot of updates in other libraries and this generates environment instability

**Comment by vale-salvatelli** [comment]
No need to revert - I was just trying to understand.

**Review by vale-salvatelli — APPROVED** [review]
